### PR TITLE
Message Detail in SwiftUI

### PR DESCRIPTION
**Pull request by MartinJibber**
Message reading functionality isn't implemented in this PR so I'm merging into a separate feature branch.

Also, this PR isn't as big as it looks. I renamed some things which is why the line count is so high.

### CODE REVIEW

File: Shared/Models/Message Model/Messageable.swift [hunk 1]
```diff
@@ -23,7 +23,7 @@ protocol Messageable {
     var isFromCurrentUser: Bool { get }
     var authorId: String { get }
     var attributes: [String: Any]? { get }
-    var avatar: Avatar { get }
+    var avatar: Avatar? { get }
```
Comment (MartinJibber) [inline Shared/Models/Message Model/Messageable.swift:26]: I made a mock message object for SwiftUI testing. The mock doesn't have an avatar to show so this is optional.

File: Shared/Models/Message Model/MockMessage.swift [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+//  MockMessage.swift
+//  Jibber
+//
+//  Created by Martin Young on 2/8/22.
+//  Copyright © 2022 Benjamin Dodgson. All rights reserved.
+//
+
+import Foundation
+
+struct MockMessage: Messageable {
```
Comment (MartinJibber) [inline Shared/Models/Message Model/MockMessage.swift:11]: Mock message for testing.

File: Shared/UI/Views/MessageContentView/MessagePreviewViewController.swift [hunk 3]
```diff
@@ -13,9 +13,9 @@ class MessagePreviewViewController: ViewController {
     let message: Messageable
 
     private let content = MessageContentView()
-    private let readView = MessageReadView()
-    private let emotionView = EmotionView()
-    private let replyView = MessageReplyView()
+    private let readView = old_MessageReadView()
+    private let emotionView = old_EmotionView()
+    private let replyView = old_MessageReplyView()
```
Comment (MartinJibber) [inline Shared/UI/Views/MessageContentView/MessagePreviewViewController.swift:18]: The new message detail views are now first class citizens while the old ones are getting deprecated.

File: iOS/Extensions/SwiftUI/Shape+Extensions.swift [hunk 4]
```diff
@@ -18,4 +18,12 @@ extension Shape {
...
         self.stroke(Color(color.color.withAlphaComponent(alpha)), lineWidth: width)
     }
+
+    func fill(color: ThemeColor,
+              strokeColor: ThemeColor,
+              lineWidth: CGFloat) -> some View {
+
+        self.stroke(Color(color.color), lineWidth: lineWidth)
+            .background(self.fill(Color(color.color)))
+    }
```
Comment (MartinJibber) [inline iOS/Extensions/SwiftUI/Shape+Extensions.swift:28]: SwiftUI convenience function to set color and stroke color in one call.

File: iOS/Flows/Conversation/Cells/MessageCell.swift [hunk 5]
```diff
@@ -13,7 +13,11 @@ import SwiftUI
 class MessageCell: UICollectionViewCell {
 
     let content = MessageContentView()
-    let detailView = MessageDetailView()
+
+    // Detail View
+    @ObservedObject private var detailState = MessageDetailConfig(message: nil)
+    private lazy var detailView = MessageDetailView(config: self.detailState)
+    private lazy var detailVC = NavBarIgnoringHostingController(rootView: self.detailView)
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/MessageCell.swift]: The detail state is basically an observable wrapper around a message. All the detail subviews have a reference to it so they can be updated when the message changes.

Also note I'm using our custom Hosting Controller subclass to get around that nav bar issue.

File: iOS/Flows/Conversation/Cells/MessageCell.swift [hunk 6]
```diff
@@ -28,40 +32,49 @@ class MessageCell: UICollectionViewCell {
...
+        self.detailVC.removeFromParentAndSuperviewIfNeeded()
+    }
+
+    override func didMoveToWindow() {
+        super.didMoveToWindow()
+
+        if self.window.exists {
+            self.parentViewController()?.addChild(viewController: self.detailVC,
+                                                  toView: self.contentView)
+        }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Cells/MessageCell.swift:50]: Making sure that the hosting controller for the detail view gets cleaned up.

File: iOS/Flows/Conversation/Status View/old_MessageDetailView.swift [hunk 7]
```diff
@@ -10,12 +10,12 @@ import Foundation
 import StreamChat
 import Combine
 
-class MessageDetailView: BaseView {
+class old_MessageDetailView: BaseView {
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Status View/old_MessageDetailView.swift:13]: Deprecated.

File: iOS/UI/SwiftUI/NavBarIgnoringHostingController.swift [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+//  Copyright © 2022 Benjamin Dodgson. All rights reserved.
+//
+
+import Foundation
+import SwiftUI
+
+/// A subclass of UIHostingController  that doesn't allow its embedded SwiftUI views to affect the UINavigationControllers nav bar visibility.
+/// NOTE: This class is a HACK that  exists to work around undesirable interactions between UINavigationControllers and embedded SwiftUI views.
+/// For some reason, embedded SwiftUI Views set the their containing NavController's nav bar to visible regardless of the navigationBarHidden state.
+class NavBarIgnoringHostingController<Content>: UIHostingController<Content> where Content: View {
```
Comment (MartinJibber) [inline iOS/UI/SwiftUI/NavBarIgnoringHostingController.swift:15]: I don't love doing this, but Apple has given me no choice.

File: iOS/UI/SwiftUI/NavBarIgnoringHostingController.swift [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+        super.viewWillAppear(animated)
+
+        // If the SwiftUI view changed the NavBar's visibility, reset it back to what it was.
+        if let wasNavigationBarHidden = wasNavigationBarHidden,
+            self.navigationController?.isNavigationBarHidden != wasNavigationBarHidden {
+
+            self.navigationController?.isNavigationBarHidden = wasNavigationBarHidden
+        }
+    }
```
Comment (MartinJibber) [inline iOS/UI/SwiftUI/NavBarIgnoringHostingController.swift:34]: I don't love doing this, but Apple has given me no choice.

File: iOS/_Stream Temp/ChatMessage+Extensions.swift [hunk 10]
```diff
@@ -84,6 +88,25 @@ extension Message: Messageable {
...
+            return nil
+        }
+
+        guard case .string(let emotionString) = emotionJSON,
+              let emotion = Emotion(rawValue: emotionString) else {
+                  return nil
+              }
+
+        return emotion
+    }
```
Comment (MartinJibber) [inline iOS/_Stream Temp/ChatMessage+Extensions.swift:108]: Convenience variable for getting the emotion off a message.

File: Shared/Extensions/UI Extensions/UIViewController+Extensions.swift [hunk 11]
```diff
@@ -19,7 +19,12 @@ extension UIViewController {
         viewController.didMove(toParent: self)
     }
 
-    func removeFromParentSuperview() {
+    func removeFromParentAndSuperviewIfNeeded() {
+        guard self.parent.exists else {
+            self.view.removeFromSuperview()
+            return
+        }
```
Comment (MartinJibber) [inline Shared/Extensions/UI Extensions/UIViewController+Extensions.swift:26]: Adding "if needed" so we don't need to check if the view controller actually has a parent.

File: iOS/_Stream Temp/ChatMessage+Extensions.swift [hunk 12]
```diff
@@ -46,15 +46,19 @@ extension Message: Messageable {
         return .passive
     }
     
-    var lastReadAt: Date? {
+    var lastUpdatedAt: Date? {
```
Comment (MartinJibber) [inline iOS/_Stream Temp/ChatMessage+Extensions.swift:49]: Change this back to last read. And just use an if statement for determining which date to show.

**Review by Benjamindodgson — APPROVED** [review]

**Pull request merged**